src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs

```diff
@@ -1,442 +1,199 @@
+        /// <typeparam name="TAttribute">type of binding attribute</typeparam>
+        /// <typeparam name="TUserType"></typeparam>
+        /// <param name="buildFromAttribute">builder function to create the object that will get passed to the user function.</param>
+        /// <returns></returns>
+        public IBindingProvider BindToExactType<TAttribute, TUserType>(Func<TAttribute, TUserType> buildFromAttribute)
+            where TAttribute : Attribute, new()
         {
-            return Activator.CreateInstance(typeof(TCollector), userContext) as IAsyncCollector<TCore>;
+            var rule = new GenericExactTypeBinder<TAttribute, TUserType>(_nameResolver, buildFromAttribute);
+            return rule;
         }
 
         /// <summary>
         /// Bind a  parameter to an IAsyncCollector. Use this for things that have discrete output items (like sending messages or writing table rows)
         /// This will add additional adapters to connect the user's parameter type to an IAsyncCollector. 
         /// </summary>
-        /// <typeparam name="TMessage">The native type of message. For example, for azure queues, this would be CloudQueueMessage.</typeparam>
-        /// <typeparam name="TContext">helper object to pass to the binding the configuration state. This can point back to context like secrets, configuration, etc.</typeparam>
-        /// <param name="parameter">parameter being bound.</param>
-        /// <param name="converterManager"></param>
-        /// <param name="builder">function to create a new instance of the underlying Collector object to pass to the parameter. 
-        /// This binder will wrap that in any adapters to make it match the requested parameter type.</param>
-        /// <param name="invokeString">a short string summary that describes this binding. That will display on the dashboard</param>
-        /// <param name="invokeStringBinder">retrieve a context from the invoke string. </param>
+        /// <typeparam name="TMessage"></typeparam>
+        /// <typeparam name="TTriggerValue"></typeparam>
+        /// <param name="bindingStrategy">a strategy object that describes how to do the binding</param>
+        /// <param name="parameter">the user's parameter being bound to</param>
+        /// <param name="converterManager">the converter manager, used to convert between the user parameter's type and the underlying native types used by the trigger strategy</param>
+        /// <param name="createListener">a function to create the underlying listener for this parameter</param>
         /// <returns></returns>
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1720:IdentifiersShouldNotContainTypeNames", MessageId = "string")]
-        public static IBinding BindCollector<TMessage, TContext>(
+        public static ITriggerBinding GetTriggerBinding<TMessage, TTriggerValue>(
+            ITriggerBindingStrategy<TMessage, TTriggerValue> bindingStrategy,
```

Comment:
I'll focus on the trigger separately. This change doesn't actually change triggers. 

---
In reply to: [61138541](https://github.com/Azure/azure-webjobs-sdk/pull/694#discussion_r61138541) [](ancestors = 61138541)